src/rust/ide/src/model/module.rs

```diff
@@ -503,6 +520,15 @@ pub trait API:Debug+model::undo_redo::Aware {
     fn with_node_metadata
     (&self, id:ast::Id, fun:Box<dyn FnOnce(&mut NodeMetadata) + '_>) -> FallibleResult;
 
+    /// Access project's metadata with a given function.
+    ///
+    /// Fails, if the project's metadata are not set in this module.
+    fn with_project_metadata_internal(&self, fun:Box<dyn FnOnce(&ProjectMetadata) + '_>);
+
+    /// Borrow mutably the project's metadata and update it with a given function.
+    fn update_project_metadata_internal
+    (&self, fun:Box<dyn FnOnce(&mut ProjectMetadata) + '_>) -> FallibleResult;
+
```

Comment:
Why something called "internal" is defined in API trait?

Reply:
Ok, a little later I saw, however the docs should have an explanation.

Reply:
Additionally, calling callback while borrowed is something I would avoid as much as possible. Perhaps we could do the same as in nodes, where we move-out the node metadata pass a reference to callback and put them back.

Reply:
I'm afraid that temporarily moving metadata out might be just as bad: if a callback calls something that will try to access metadata again, it won't find anything. It would be replacing hard panic with a invalid data (we could do the same on top of `try_borrow`). 

TBH at this point I'm beginning to think that the better API might be just getting/setting metadata by copying it. If it parses in sane time, than copying should be fast enough. However, I'd leave this change for future, as the node metadata function should be adjusted as well.

Reply:
Documentation is extended.

Reply:
Also changed name to better describe thing.